File: GigRadio/Models/SoundCloud/SoundCloudUser.m
//
//  SoundCloudUser.m
//  GigRadio
//
//  Created by Michael Forrest on 16/08/2014.
//  Copyright (c) 2014 Good To Hear. All rights reserved.
//

#import "SoundCloudUser.h"

@implementation SoundCloudUser

// Specify default values for properties

+ (NSDictionary *)defaultPropertyValues
{
    return @{
            @"website":@"",
            @"website_title":@"",
            @"permalink":@"",
            @"description":@"",
            @"userDescription":@"",
            @"avatar_url":@"",
            @"city":@""
             };
}

// Specify properties to ignore (Realm won't persist these)

//+ (NSArray *)ignoredProperties
//{
//    return @[];
//}
+(SoundCloudUser *)findById:(NSInteger)identifier{
   return [[SoundCloudUser objectsWhere:@"id == %i",identifier ] firstObject];
}

-(void)loadImage:(void (^)(UIImage *))callback{
    NSURLSession * session = [NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]];
    NSString * url = [self.avatar_url stringByReplacingOccurrencesOfString:@"-large.jpg" withString:@"-t500x500.jpg"];
    [[session dataTaskWithURL:[NSURL URLWithString:url] completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
        if(data){
            UIImage * image = [UIImage imageWithData:data];
            dispatch_async(dispatch_get_main_queue(), ^{
                callback(image);
            });
        }
    }] resume];

}

-(SoundCloudTrack *)nextTrackAfter:(NSArray *)trackIdsToSkip{
    for (SoundCloudTrack * track in self.tracks) {
        if([trackIdsToSkip indexOfObject:@(track.id)] == NSNotFound){
            return track;
        }
    }
    return nil;
}

@end

File: GigRadio/Models/SoundCloud/SoundCloudUser.swift
//
//  SoundCloudUser.swift
//  GigRadio
//
//  Created by Michael Forrest on 17/05/2015.
//  Copyright (c) 2015 Good To Hear. All rights reserved.
//

import UIKit
import RealmSwift

class SoundCloudUser: Object {
    dynamic var id: Int = 0
    dynamic var kind = ""
    dynamic var permalink = ""
    dynamic var username = ""
    dynamic var uri = ""
    dynamic var permalinkUrl = ""
    dynamic var avatarUrl = ""
    dynamic var fullName = ""
    dynamic var descriptionText  = ""
    dynamic var city = ""
    dynamic var country = ""
    dynamic var website = ""
    dynamic var websiteTitle = ""
    dynamic var trackCount: Int = 0
    dynamic var playlistCount: Int = 0
    dynamic var followersCount: Int = 0
    dynamic var followingsCount: Int = 0

    dynamic var tracks = List<SoundCloudTrack>()
    dynamic var tracksHaveBeenChecked = false
    
    override static func primaryKey()->String?{
        return "id"
    }
    var couldHaveTracksAvailable:Bool{
        return tracksHaveBeenChecked == false || tracks.count > 0
    }
    func displayName()->String{
       var result = fullName
        if result == ""{
            result = username
        }
        return result
    }
    func showOnSoundCloud(){
        let url = NSURL(string:"soundcloud://users:\(id)")!
        let app = UIApplication.sharedApplication()
        if app.canOpenURL(url){
            app.openURL(url)
        }else{
            let url = NSURL(string: "https://soundcloud.com/\(permalink)")!
            app.openURL(url)
        }
    }
}
